### PR TITLE
[WIP] Update to Unicode 16.0.0

### LINKED ISSUE #77: Update unicode
I attempted to update to 15.1.0 but looks like the codegen will need an update:

```
build/unicode_gen unicode libunicode-table.h
Property not found: InCB;
```

### LINKED ISSUE #530: Unicode needs to be updated to 16.0.0
https://www.unicode.org/versions/Unicode16.0.0/

### PR DESCRIPTION
**Pull request by linusg**
Please review this very carefully, I have no idea what I'm doing :)

Closes #77.
Closes #530.

### CODE REVIEW

File: libunicode.c [hunk 1]
```diff
@@ -107,11 +112,16 @@ static int lre_case_conv_entry(uint32_t *res, uint32_t c, int conv_type, uint32_
...
     case RUN_TYPE_U_EXT:
+    case RUN_TYPE_L_EXT:
     case RUN_TYPE_LF_EXT:
         if (is_lower != (type - RUN_TYPE_U_EXT))
             break;
         c = case_conv_ext[data];
         break;
+    case RUN_TYPE_U_EXT2:
+    case RUN_TYPE_U_EXT3:
+        // TODO
```
Comment (linusg) [inline libunicode.c:123]: Someone needs to tell me how to implement these, my brain is not capable of writing this kind of obfuscated C code.

File: unicode_gen.c [hunk 2]
```diff
@@ -1019,6 +1023,27 @@ void find_run_type(TableEntry *te, CCInfo *tab, int code)
...
+                len++;
+            }
+            te->len = len;
+            te->type = RUN_TYPE_U2_399_EXT2;
+            te->ext_data[0] = ci->u_data[0];
+            te->ext_data[1] = ci->l_data[0];
+            te->ext_len = 2;
+            return;
+        }
+
```
Comment (linusg) [inline unicode_gen.c:1046]: Copy and pasted from the one below except there is no `F` component

File: libunicode.c (hunk 1 above)
Comment (bnoordhuis) [reply libunicode.c:123]: Caveat emptor, I'm not 100% sure but I suspect you can merge the `RUN_TYPE_U_EXT2` case with `RUN_TYPE_LF_EXT2` and update the guard to:
```c
if (is_lower != (type - RUN_TYPE_U_EXT2))
    break;
```
`RUN_TYPE_U_EXT2` should immediately precede `RUN_TYPE_LF_EXT2` in the enum declaration for that to work (not the case right now.)

For `RUN_TYPE_U_EXT3`, the code should probably look like this:
```c
case RUN_TYPE_U_EXT3:
    res[0] = c - code + case_conv_ext[data >> 8];
    res[1] = case_conv_ext[(data >> 4) & 0x0f];
    res[2] = case_conv_ext[data & 0x0f];
    return 3;
```

File: unicode_gen.c (hunk 2 above)
Comment (bnoordhuis) [reply unicode_gen.c:1046]: That means the if statement below is dead code now, right? Less specific subsumes more specific.

It's probably okay if you swap them. The code itself looks correct at a superficial glance.

File: unicode_gen.c [hunk 3]
```diff
@@ -1239,6 +1277,11 @@ void build_conv_table(CCInfo *tab)
                     te->ext_len = 1;
                     te->ext_data[0] = te->data;
                     break;
+                case RUN_TYPE_L:
+                    te->type = RUN_TYPE_L_EXT;
+                    te->ext_len = 1;
+                    te->ext_data[0] = te->data;
+                    break;
```
Comment (bnoordhuis) [inline unicode_gen.c:1284]: With some clever rearranging of the enum values, you can avoid code duplication by folding the three cases into one:
```c
case RUN_TYPE_L:
case RUN_TYPE_LF:
case RUN_TYPE_U:
    te->type += RUN_TYPE_L_EXT - RUN_TYPE_L;
    te->ext_len = 1;
    te->ext_data[0] = te->data;
    break;
```

**Comment by bnoordhuis** [comment]
@linusg do you plan on revisiting this?

FWIW, I don't think this fully closes out #77 until the regex engine is taught about unicode emoji ([tr51](https://unicode.org/reports/tr51/)). Ex. `/\p{Basic_Emoji}/` with the challenge that emoji are variable-length encodings.

I did a proof of concept where I pregenerate the regex bytecode for Basic_Emoji, Emoji_Keycap_Sequence, etc. and memcpy it when compiling regular expressions but that results in _really_ bloated bytecode.

**Comment by linusg** [comment]
I won't get around to it until sometime next week at the earliest, if someone else wants to pick this up before then that's fine by me :)

**Comment by bnoordhuis** [comment]
No rush. I was just wondering if I should adopt it.

**Comment by linusg** [comment]
I didn't find the time for this yet and likely won't in the coming weeks either - if you or someone else wants to finish it please feel free, I'd love to have this merged!

**Comment by bnoordhuis** [comment]
I took a slightly different tack because I wanted to understand why 15.1 behaved differently from 16.0 but the update to 15.1 was done in 416ab66f0bf3cb8b5250b478bb78f812901f8be6, followed immediately afterwards by an update to 16.0 in 243b968ee12068b721ca2613ff3d0e01ab5dc2ea. Thanks anyway for the pull request!

**Comment by linusg** [comment]
Thank you, much appreciated!

**Pull request closed without merging**